program/src/processor.rs

```diff
@@ -579,8 +611,42 @@ pub fn withdraw_excess_inactive_sol<'a, 'b>(
     Ok(excess_balance)
 }
 
-/// Recover any inactive balance from the validator's stake accounts to the
-/// reserve account.
+pub fn compare_stake_account<'a, 'b>(
+    account_iter: &'a mut Iter<AccountInfo<'b>>,
+    n_unstake_accounts: usize,
+    n_stake_accounts: usize,
+    seed: u64,
+    stake_account_address: &Pubkey,
+    accounts: &'a WithdrawInactiveStakeInfo<'a, 'b>,
+    stake_type: StakeType,
+) -> Result<&'a AccountInfo<'b>, LidoError> {
+    match account_iter.next() {
+        None => {
+            msg!(
+                "Not enough {} accounts provided, got {} but expected {}.",
+                stake_type,
+                accounts.stake_accounts.len() - n_unstake_accounts as usize,
```

Comment:
This can overflow, for example, if you expect 1 stake account and 1 unstake account, but provide zero accounts. Then here it will do 0 - 1. I’m not sure if that causes a crash or a wraparound when compiled to Solana BPF.

We might move this check into `process_withdraw_inactive_stake`, something like
```rust
if accounts.stake_accounts.len() != n_stake_accounts + n_unstake_accounts {
    msg!("Wrong number of stake accounts provided, expected {} stake accounts and {} unstake accounts, \
        but got {} accounts.", n_stake_accounts, n_unstake_accounts, accounts.stake_accounts.len(),
    );
    return Err(LidoError::InvalidStakeAccount);
}

// Does not panic, because len = n_stake_accounts + n_unstake_accounts >= n_stake_accounts.
let (stake_accounts, unstake_accounts) = accounts.stake_accounts.split_at(n_stake_accounts);
```

Then we can zip the seeds with the accounts and do
```rust
for (seed, stake_accounts) in validator.entry.stake_seeds.zip(&stake_accounts) {
    let (stake_account_address, _bump_seed) = validator.find_stake_account_address(program_id, accounts.lido.key, seed);
    if account.key != stake_account_address {
        msg!("Wrong stake account for seed {}: expected {} but got {}.", seed, stake_account_address, account.key);
        return Err(LidoError::InvalidStakeAccount);
     }
}
// After this it is safe to use `account`.
```

It’s about the same size as the call to `compare_stake_account` is now. It’s a bit more duplication, but I think it actually makes things clearer because we don’t have to handle the two slightly different cases in one function.